lib/explorer/series.ex

```diff
@@ -3167,10 +3171,20 @@ defmodule Explorer.Series do
     end
   end
 
-  # TODO: fix the logic for integer dtypes
-  defp cast_to_add({:s, left}, {:s, right}), do: {:s, max(left, right)}
-  defp cast_to_add({:s, _}, {:f, _} = float), do: float
-  defp cast_to_add({:f, _} = float, {:s, _}), do: float
+  # Review the size needed for this operation.
+  defp cast_to_add({int_type, left}, {int_type, right}) when K.in(int_type, [:s, :u]),
+    do: {int_type, max(left, right)}
+
+  defp cast_to_add({:s, s_size}, {:u, u_size}) when u_size >= s_size,
+    do: {:s, min(64, u_size * 2)}
+
+  defp cast_to_add({:u, u_size}, {:s, s_size}) when u_size >= s_size,
+    do: {:s, min(64, u_size * 2)}
+
+  defp cast_to_add({:s, s_size}, {:u, _}), do: {:s, s_size}
+  defp cast_to_add({:u, _}, {:s, s_size}), do: {:s, s_size}
+  defp cast_to_add({int_type, _}, {:f, _} = float) when K.in(int_type, [:s, :u]), do: float
+  defp cast_to_add({:f, _} = float, {int_type, _}) when K.in(int_type, [:s, :u]), do: float
   defp cast_to_add({:f, _}, {:f, _}), do: {:f, 64}
```

Comment:
Merging between floats should be the max of their size. :)

Reply:
In Nx, the logic for merging types for +, - and * are all the same. So maybe we should add a function called `cast_numeric` that holds this shared logic. This way, we can write `cast_to_add` like this:

```
  defp cast_to_add(:date, {:duration, _}), do: :date
  defp cast_to_add({:duration, _}, :date), do: :date
  defp cast_to_add({:datetime, p}, {:duration, p}), do: {:datetime, p}
  defp cast_to_add({:duration, p}, {:datetime, p}), do: {:datetime, p}
  defp cast_to_add({:duration, p}, {:duration, p}), do: {:duration, p}
  defp cast_to_add(left, right), do: cast_numeric(left, right)
```

?

The same for `-` and `*`?

Reply:
Wait, maybe this logic should be added to `cast_for_arithmetic` instead? 🤔 That's where we already cast both sides. And then this function just check if both sides are the same? And then we likely don't need the changes in PolarsBackend.Series.subtract?

Reply:
I will check that! Currently these functions have different roles: `cast_for_arithmetic` really cast the series before perform the operation (at the Series module level), and `cast_to_add/subtract/divide/*` does not cast yet, but calculates the `out_dtype` so the backend (in this case `PolarsBackend.Series` or the lazy series) can cast.

Reply:
@josevalim I made the simplification in 566d9772216f089a728c23dd89b09816c1778893. I'm going to try that other change we talked soon.